lib/srv/db/mysql/handshake_test.go

```diff
@@ -0,0 +1,290 @@
+		wantHandshake      bool
+	}{
+		{
+			agentHandshakeMode: handshakeDisabled,
+			proxyHandshakeMode: handshakeDisabled,
+		},
+		{
+			agentHandshakeMode: handshakeWhenSupported,
+			proxyHandshakeMode: handshakeDisabled,
+		},
+		{
+			agentHandshakeMode: handshakeDisabled,
+			proxyHandshakeMode: handshakeWhenSupported,
+		},
+		{
+			agentHandshakeMode: handshakeWhenSupported,
+			proxyHandshakeMode: handshakeWhenSupported,
+			wantHandshake:      true,
+		},
+		{
+			agentHandshakeMode: handshakeAlways,
+			proxyHandshakeMode: handshakeWhenSupported,
+			wantHandshake:      true,
+		},
+		{
+			agentHandshakeMode: handshakeAlways,
+			proxyHandshakeMode: handshakeAlways,
+			wantHandshake:      true,
+		},
+		// always and disabled are not compatible
+		{
+			agentHandshakeMode: handshakeAlways,
+			wantAgentError:     "EOF",
+			proxyHandshakeMode: handshakeDisabled,
+			wantProxyError:     "expected OK or ERR packet",
+		},
+		{
+			agentHandshakeMode: handshakeDisabled,
+			proxyHandshakeMode: handshakeAlways,
+			wantProxyError:     "invalid protocol",
+		},
+		// proxy has to be upgraded last, it cant be changed to always
+		// handshake until all supported agents will always handshake
+		{
+			agentHandshakeMode: handshakeWhenSupported,
+			proxyHandshakeMode: handshakeAlways,
+			wantProxyError:     "invalid protocol",
+		},
+	}
+	for _, test := range tests {
+		ctx := t.Context()
+		t.Run(fmt.Sprintf("agent handshake %v proxy handshake %v", test.agentHandshakeMode, test.proxyHandshakeMode), func(t *testing.T) {
+			t.Parallel()
+			agentProxyConn, proxyAgentConn := net.Pipe()
+			var wg sync.WaitGroup
+			wg.Add(2)
+
+			// agent
+			var agentMySQLServer *server.Conn
+			go func() {
+				defer wg.Done()
+				defer agentProxyConn.Close()
+				agentMySQLServer = makeProxyConn(agentProxyConn)
+				log := slog.With(teleport.ComponentKey, teleport.ComponentDatabase)
+				err := notifyProxy(ctx, log, test.agentHandshakeMode, agentMySQLServer)
+				if test.wantAgentError != "" {
+					assert.ErrorContains(t, err, test.wantAgentError)
+					return
+				}
+				if !assert.NoError(t, err) {
+					return
+				}
+				if test.wantHandshake {
+					assert.NotEmpty(t, agentMySQLServer.Attributes())
+					assert.Equal(t, "llama", agentMySQLServer.Attributes()["species"])
+				} else {
+					assert.Empty(t, agentMySQLServer.Attributes())
+				}
+				assert.Zero(t, agentMySQLServer.Sequence)
+				assertEmptyConn(t, agentMySQLServer.Conn)
+			}()
+
+			// proxy
+			go func() {
+				defer wg.Done()
+				defer proxyAgentConn.Close()
+				log := slog.With(teleport.ComponentKey, teleport.ComponentProxy)
+				err := waitForAgent(ctx, log, test.proxyHandshakeMode, proxyAgentConn, func(c *client.Conn) error {
+					c.SetAttributes(map[string]string{"species": "llama"})
+					return nil
+				})
+				if test.wantProxyError != "" {
+					assert.ErrorContains(t, err, test.wantProxyError)
+					return
+				}
+				if !assert.NoError(t, err) {
+					return
+				}
```

Comment:
this is totally fine. alternatively you can check out these outside the goroutine.